crates/sp-domains-fraud-proof/src/fraud_proof.rs

```diff
@@ -288,52 +291,117 @@ pub enum VerificationError {
-    bundle_index: u32,
+pub struct FalseInvalidBundleEntryFraudProof {
+    pub bad_receipt_hash: ReceiptHash,
+    pub domain_id: DomainId,
+    pub bundle_index: u32,
 }
 
-impl MissingInvalidBundleEntryFraudProof {
-    pub fn new(domain_id: DomainId, bundle_index: u32) -> Self {
+impl FalseInvalidBundleEntryFraudProof {
+    pub fn new(bad_receipt_hash: ReceiptHash, domain_id: DomainId, bundle_index: u32) -> Self {
         Self {
+            bad_receipt_hash,
             domain_id,
             bundle_index,
         }
     }
 }
 
-// TODO: Define rest of the fraud proof fields
+/// Proof data specific to each *expected* invalid bundle type
+#[derive(Debug, Decode, Encode, TypeInfo, PartialEq, Eq, Clone)]
+pub enum ProofDataPerExpectedInvalidBundle {
+    OutOfRangeTx,
+}
+
 #[derive(Debug, Decode, Encode, TypeInfo, PartialEq, Eq, Clone)]
-pub struct ValidAsInvalidBundleEntryFraudProof {
-    domain_id: DomainId,
-    bundle_index: u32,
+pub struct TrueInvalidBundleEntryFraudProof {
+    pub bad_receipt_hash: ReceiptHash,
+    pub domain_id: DomainId,
+    pub bundle_index: u32,
+    pub mismatched_extrinsic_index: u32,
+    pub extrinsic_inclusion_proof: Vec<Vec<u8>>,
+    pub proof_data: ProofDataPerExpectedInvalidBundle,
 }
 
-impl ValidAsInvalidBundleEntryFraudProof {
-    pub fn new(domain_id: DomainId, bundle_index: u32) -> Self {
+impl TrueInvalidBundleEntryFraudProof {
+    pub fn new(
+        bad_receipt_hash: ReceiptHash,
+        domain_id: DomainId,
+        bundle_index: u32,
+        mismatched_extrinsic_index: u32,
+        extrinsic_inclusion_proof: Vec<Vec<u8>>,
+        proof_data: ProofDataPerExpectedInvalidBundle,
+    ) -> Self {
         Self {
+            bad_receipt_hash,
             domain_id,
             bundle_index,
+            mismatched_extrinsic_index,
+            extrinsic_inclusion_proof,
+            proof_data,
+        }
+    }
+
+    pub fn matches_with_bundle_entry(&self, inboxed_bundle_entry: &InboxedBundle) -> bool {
+        if !inboxed_bundle_entry.is_invalid() {
+            return true;
         }
+
+        let expected_bundle_validity = match self.proof_data {
+            ProofDataPerExpectedInvalidBundle::OutOfRangeTx => BundleValidity::Invalid(
+                InvalidBundleType::OutOfRangeTx(self.mismatched_extrinsic_index),
+            ),
+        };
+
+        expected_bundle_validity == inboxed_bundle_entry.bundle
     }
 }
 
-/// Fraud proof indicating that `invalid_bundles` field of the receipt is incorrect
+/// Fraud proof indicating that a bundle included in `inboxed_bundles` field has incorrect
+/// `bundle_validity` field
 #[derive(Debug, Decode, Encode, TypeInfo, PartialEq, Eq, Clone)]
 pub enum InvalidBundlesFraudProof {
-    MissingInvalidBundleEntry(MissingInvalidBundleEntryFraudProof),
-    ValidAsInvalid(ValidAsInvalidBundleEntryFraudProof),
+    TrueInvalid(TrueInvalidBundleEntryFraudProof),
+    FalseInvalid(FalseInvalidBundleEntryFraudProof),
```

Comment:
I wonder if we need 2 separate structs here, the only difference between `TrueInvalid` and `FalseInvalid` should be just the last check, i.e. `if is_tx_in_range` vs `if !is_tx_in_range`, so simply a flag can handle this already.